File: src/abstract/scalar_with_id.ts
import _ from "underscore";

export interface ObjectWithIdAndValue<T> {
    id: number;
    value: T;
}

export type ValueOrObject<T> = ObjectWithIdAndValue<T> | T;

export type ValueOrObjectArray<T> = ObjectWithIdAndValue<T>[] | T[];

export function isObjectWithIdAndValue<T>(
    valueOrObject: ValueOrObject<T>,
): valueOrObject is ObjectWithIdAndValue<T> {
    return Boolean(_.isObject(valueOrObject) && !_.isArray(valueOrObject) && valueOrObject.value);
}

/**
 * Helper class representing a scalar with an associated id.
 */
export class ScalarWithId<T> {
    id: number;

    value: T;

    /**
     * Create a an array with ids.
     * @param valueOrObject - a ScalarWithID, or any other type.
     * @param id - numerical id (Integer).
     */
    constructor(valueOrObject: ValueOrObject<T>, id = 0) {
        // if already passing a ScalarWithId => preserve original
        if (isObjectWithIdAndValue(valueOrObject)) {
            // NOTE - Arrays are Objects too
            this.id = valueOrObject.id;
            this.value = valueOrObject.value;
        } else {
            this.id = id;
            this.value = valueOrObject;
        }
    }

    /**
     * Serialize class instance to JSON.
     * @example {"id" : 0, "value" : "Si" }
     */
    toJSON(): ObjectWithIdAndValue<T> {
        return {
            id: this.id,
            value: this.value,
        };
    }
}

File: src/py/mat3ra/made/abstract/scalar_with_id.py
from typing import Union, List, TypeVar, Generic, Dict

T = TypeVar('T')


class ObjectWithIdAndValue(Generic[T]):
    def __init__(self, id: int, value: T):
        self.id = id
        self.value = value


ValueOrObject = Union[ObjectWithIdAndValue[T], T]
ValueOrObjectArray = Union[List[ObjectWithIdAndValue[T]], List[T]]


def is_object_with_id_and_value(value_or_object: ValueOrObject[T]) -> bool:
    return isinstance(value_or_object, ObjectWithIdAndValue)


class ScalarWithId(Generic[T]):
    def __init__(self, value_or_object: ValueOrObject[T], id: int = 0):
        if is_object_with_id_and_value(value_or_object):
            self.id = value_or_object.id
            self.value = value_or_object.value
        else:
            self.id = id
            self.value = value_or_object

    def to_json(self) -> Dict[str, Union[int, T]]:
        return {
            "id": self.id,
            "value": self.value
        }
